README.md

```diff
@@ -337,6 +339,190 @@ PING 10.0.0.1 (10.0.0.1) 56(84) bytes of data.
 rtt min/avg/max/mdev = 0.054/0.141/0.342/0.117 ms
 ```
 
+## Testing environment (virtio)
+Below is our testing environment with virtio feature:
+
+<p align="center"><img src="assets/vwifi_virtio_arch.drawio.png" alt="vwifi virtio image" width=60%></p>
+
+vwifi makes use of the `virtio-net` device, Linux tap device, and Linux bridge device to fulfill inter-guest communication.
+In our testing environment, on the host side, we create three `tap` devices and one `bridge` device, then attach each `tap` device to the `bridge` device. And we create three VMs by `Qemu` with the `virtio-net` devices and specify that every `virtio-net` device connects to a `tap` device in the host.
+In VM1, the network interface created by `vwifi` will be in HostAP mode, with the user program `hostapd` running on top of it.
+In VM2 and VM3, we have an STA mode interface and `wpa_supplicant` running on top of it.
+## Build and Run (virtio)
+Below describes how to build a minimal workable VM environment.
+### Build Linux kernel
+Download Linux kernel source from [kernel.org](https://kernel.org/). 
+
+Enter the top directory of the Linux kernel source and use the following command to generate the configuration file:
+```shell
+make menuconfig
+```
+The default kernel cofiguration will work for our testing environment, so just click `save` and we get `.config` on the top directory.
+
+Before building the kernel, please ensure that all the needed packges or libraries have been installed in your machine.
+
+Then Build the kernel:
+```
+make -j<num threads>
+```
+
+Once we have finished building the kernel, the `bzImage` will be in the `arch/<architecture>/boot/` directory. For x86 machine, that is `arch/x86/boot/bzImage`.
+
+### Build `vwifi`
+Clone the `vwifi` repository and enter the top directory.
+
+`vwifi` let the `$(KDIR)` in `Makefile` point to the default kernel source on the host, but the kernel version of the host may differ from the one for the VMs, so we need to replace it to point to the kernel source we previously built for the VMs:
+```shell
+KDIR = <linux kernel top directory for VMs>
+```
+
+Save the change and build `vwifi`:
+```shell
+make
+```
+
+On success, we expect that `vwifi.ko` will show in the top directory of `vwifi`.
+
+If there is a compile error, please notify us by Github issue, email, or better: the patch :). The cfg80211 API and net device API in the kernel have changed along with the new kernel version, and forgive us for not testing for all of them.
```

Comment:
Remove the statement "If there is a compile error, please notify us by Github issue, email, or better: the patch"